File: ctr_mode.cpp

#include "ctr_mode.hpp"
#include "utils.hpp" // Para xor_bytes
#include <stdexcept>
#include <algorithm> // Para std::min

// Implementação do incremento do nonce (big-endian)
void increment_nonce(std::vector<uint8_t>& nonce) {
    for (int i = nonce.size() - 1; i >= 0; --i) {
        if (nonce[i] < 0xFF) {
            nonce[i]++;
            return;
        }
        nonce[i] = 0x00; // Carry
    }
    // Se chegou aqui, houve overflow em todos os bytes (raro, mas possível)
}

std::vector<uint8_t> ctr_mode_encrypt(
    const std::vector<uint8_t>& data,
    std::vector<uint8_t> nonce, // Passado por valor para modificar localmente, ou passar referência e clonar
    const void* round_keys_ptr,
    block_cipher_func_t encrypt_func,
    size_t cipher_block_size_bytes
) {
    if (nonce.size() != cipher_block_size_bytes) {
        throw std::runtime_error("Nonce size must match cipher block size.");
    }

    std::vector<uint8_t> output_data = data; // Copia os dados de entrada
    std::vector<uint8_t> counter_block = nonce;

    for (size_t i = 0; i < data.size(); i += cipher_block_size_bytes) {
        std::vector<uint8_t> encrypted_counter = encrypt_func(counter_block, round_keys_ptr);

        size_t remaining_bytes = data.size() - i;
        size_t bytes_to_xor = std::min(cipher_block_size_bytes, remaining_bytes);

        for (size_t j = 0; j < bytes_to_xor; ++j) {
            output_data[i + j] = data[i + j] ^ encrypted_counter[j];
        }
        increment_nonce(counter_block);
    }
    return output_data;
}

File: ctr_mode.py
from utils import xor_bytes

def ctr_mode_encrypt(data: bytes, nonce: bytes, key_schedule, encrypt_func, cipher_block_size_bytes: int) -> bytes:
    if len(nonce) != cipher_block_size_bytes:
        raise ValueError("Tamanho do Nonce deve ser igual ao tamanho do bloco da cifra.")
    output_data = bytearray()
    current_counter_bytes = bytearray(nonce)
    for i in range(0, len(data), cipher_block_size_bytes):
        keystream_block = encrypt_func(bytes(current_counter_bytes), key_schedule)
        data_chunk = data[i : i + cipher_block_size_bytes]
        output_data.extend(xor_bytes(data_chunk, keystream_block[:len(data_chunk)]))
        val = int.from_bytes(current_counter_bytes, 'big')
        val = (val + 1) & ((1 << (cipher_block_size_bytes * 8)) -1)
        current_counter_bytes = bytearray(val.to_bytes(cipher_block_size_bytes, 'big'))
    return bytes(output_data)
